backend/eventloop/cluster_reconciler.go

```diff
@@ -0,0 +1,206 @@
+		appName := getArgoCDApplicationName(obj.GetLabels())
+		if appName != "" && obj.GetDeletionTimestamp() == nil && !strings.HasPrefix(obj.GetNamespace(), "openshift") {
+			// Check if a GitOpsDeployment exists with the UID specified in the Application name.
+			gitopsDeplList := &managedgitopsv1alpha1.GitOpsDeploymentList{}
+			err = c.client.List(ctx, gitopsDeplList, &client.ListOptions{
+				Namespace: obj.GetNamespace(),
+			})
+			if err != nil {
+				log.Error(err, "failed to list GitOpsDeployments", "namespace", obj.GetNamespace())
+				continue
+			}
+
+			expectedUID := extractUIDFromApplicationName(appName)
+			found := false
+			for _, gitopsDepl := range gitopsDeplList.Items {
+				if gitopsDepl.UID == types.UID(expectedUID) {
+					found = true
+					break
+				}
+			}
+
+			// The resource was managed by Argo CD i.e. it has label "app.kubernetes.io/instance"
+			// But it doesn't have a corresponding GitOpsDeployment so it can be deleted.
+			if !found {
+				if err := c.client.Delete(ctx, &apiObjects[i]); err != nil {
+					log.Error(err, "failed to delete object in the orphaned reconciler", "name", obj.GetName(), "namespace", obj.GetNamespace(), "gvk", obj.GroupVersionKind())
+					continue
+				}
+
+				log.Info("Deleted an orphaned resoure that is not managed by Argo CD anymore", "Name", obj.GetName(), "Namespace", obj.GetNamespace())
+			}
+		}
+	}
+}
+
+// getAllNamespacedAPIResources returns all namespace scoped resources from a Kubernetes cluster.
+func (c *ClusterReconciler) getAllNamespacedAPIResources(ctx context.Context, log logr.Logger, opts ...client.ListOption) ([]unstructured.Unstructured, error) {
+	apiResourceList, err := c.discoveryClient.ServerPreferredNamespacedResources()
+	if err != nil {
+		return nil, err
+	}
+
+	expectedVerbs := []string{"list", "get", "delete"}
+	resources := []unstructured.Unstructured{}
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+
+	for _, apiResources := range apiResourceList {
+		for _, apiResource := range apiResources.APIResources {
+			// Ignore resources that are either cluster scoped or do not support the required verbs
+			if !apiResource.Namespaced || !isResourceAllowed(expectedVerbs, apiResource.Verbs) {
+				continue
+			}
+
+			// Since we want to list resources across all namespaces in a cluster, we can
+			// run the List API for each resource in a separate goroutine and aggregate the result.
+			wg.Add(1)
+			go func(apiResources *metav1.APIResourceList, apiResource metav1.APIResource) {
+				defer wg.Done()
+
+				objList := &unstructured.UnstructuredList{}
+				objList.SetAPIVersion(apiResources.GroupVersion)
+				objList.SetKind(apiResource.Kind)
+
+				if err := c.client.List(ctx, objList, opts...); err != nil {
+					log.V(logutil.LogLevel_Debug).Error(err, "failed to list resources", "resource", apiResource.Kind)
+					return
+				}
+
+				mu.Lock()
+				resources = append(resources, objList.Items...)
+				mu.Unlock()
+			}(apiResources, apiResource)
```

Comment:
I'm a big fan of using goroutines/concurrency to speed up slow I/O requests :+1:, BUT :grin:,  I think in this very specific case, using a goroutine per resource might be a bit too aggressive re: issuing too many requests to the K8s API server at once.

One possible option would be to change this logic to only allow X number of concurrent requests at once. (2? 3? 5?). This would be one approach to explore, if you're interested. Another option might be investigating the throttling options of the client object.

But personally, I'd be fine with just doing it the slow, steady way: using a single thread, and just steadily going through the list.

Why? In general, orphaned Argo CD resources should be a pretty rare occurrence: the vast majority of the time Argo CD will successfully clean up all the resources of the Application, I would expect.  So I think it's perfectly fine if it takes this cleanup job a little while to find those orphaned resouces in the rare case that they do appear.

WDYT?